torch/csrc/api/include/torch/optim/adagrad.h

```diff
@@ -18,40 +19,81 @@ class InputArchive;
+  ~AdagradOptions() = default;
+};
+
+struct TORCH_API AdagradParamState : public OptimizerCloneableParamState<AdagradParamState> {
+  TORCH_ARG(torch::Tensor, sum);
+  TORCH_ARG(int64_t, step);
+
+public:
+  void serialize(torch::serialize::InputArchive& archive);
+  void serialize(torch::serialize::OutputArchive& archive) const;
+  TORCH_API friend bool operator==(const AdagradParamState& lhs, const AdagradParamState& rhs);
+  ~AdagradParamState() = default;
 };
 
 class TORCH_API Adagrad : public Optimizer {
  public:
-  template <typename ParameterContainer>
+  explicit Adagrad(std::vector<OptimizerParamGroup> param_groups,
+      AdagradOptions defaults) : Optimizer(std::move(param_groups), std::make_unique<AdagradOptions>(defaults)) {
+    TORCH_CHECK(defaults.learning_rate() >= 0, "Invalid learning rate: ", defaults.learning_rate());
+    TORCH_CHECK(defaults.lr_decay() >= 0, "Invalid lr_decay value: ", defaults.lr_decay());
+    TORCH_CHECK(defaults.weight_decay() >= 0, "Invalid weight_decay value: ", defaults.weight_decay());
+    TORCH_CHECK(defaults.initial_accumulator_value() >= 0, "Invalid initial_accumulator_value value: ", defaults.initial_accumulator_value());
+    TORCH_CHECK(defaults.eps() >= 0, "Invalid epsilon value: ", defaults.eps());
+
+    for (const auto& group : param_groups_) {
+      for (const auto& p : group.params()) {
+        auto state = std::make_unique<AdagradParamState>();
+        state->step(0);
+        state->sum(torch::full_like(p.data(), defaults.initial_accumulator_value()));
+        state_[c10::guts::to_string(p.unsafeGetTensorImpl())] = std::move(state);
+      }
+    }
+  }
+
+  // TODO: we might want to replace `std::vector<Tensor>` with `ParameterContainer` at some point
   explicit Adagrad(
-      ParameterContainer&& parameters,
-      const AdagradOptions& options_)
-      : Optimizer(std::forward<ParameterContainer>(parameters)),
-        options(options_) {}
+      std::vector<Tensor> params,
+      AdagradOptions defaults) : Adagrad({std::move(OptimizerParamGroup(params))}, defaults) {}
 
   void step() override;
 
-  AdagradOptions options;
+  /// Adds the given vector of parameters to the optimizer's parameter list.
+  void add_parameters(const std::vector<Tensor>& parameters) override;
+
+  /// Provides a const reference to the parameters this optimizer holds.
+  const std::vector<Tensor>& parameters() const noexcept override;
+
+  /// Provides a reference to the parameters this optimizer holds.
+  std::vector<Tensor>& parameters() noexcept override;
+
+  /// Returns the number of parameters referenced by the optimizer.
+  size_t size() const noexcept override;
 
   void save(serialize::OutputArchive& archive) const override;
   void load(serialize::InputArchive& archive) override;
 
-  std::vector<Tensor> sum_buffers;
-  std::vector<int64_t> step_buffers;
-
  private:
-  Adagrad() : options(0) {}
+  Adagrad() {}
```

Comment:
the one we already removed? (i.e. Adagrad() : options(0) {}) or Adagrad() {} ?
We don't have options in Adagrad anymore and we have defaults_ member variable in optimizer.h which we can use for all optimizers. In case you were talking about Adagrad() {}, yeah I think we should be able to remove it since it's just a default constructor and we are not redefining it so it should be fine